src/mito2/src/read/compat.rs

```diff
@@ -239,9 +254,10 @@ fn may_compat_fields(
     mapper: &ProjectionMapper,
     actual: &RegionMetadata,
 ) -> Result<Option<CompatFields>> {
+    let expect = mapper.metadata();
     let expect_fields = mapper.batch_fields();
     let actual_fields = Batch::projected_fields(actual, mapper.column_ids());
-    if expect_fields == actual_fields {
+    if expect.schema == actual.schema {
```

Comment:
If we only read a subset of fields (projection) then this check will also return false. We can returns the column id and data type in `batch_fields()` and `projected_fields()` above.